File: native-apps/desktop/windows/src/PetroNetDesktop.Domain/Primitives/ApiError.cs
namespace PetroNetDesktop.Domain.Primitives;

public sealed record ApiError(ApiErrorCode Code, string Message, string? Detail = null)
{
    public static ApiError Network(string message, string? detail = null) =>
        new(ApiErrorCode.Network, message, detail);

    public static ApiError Unauthorized(string message, string? detail = null) =>
        new(ApiErrorCode.Unauthorized, message, detail);

    public static ApiError CertificatePinFailure(string message, string? detail = null) =>
        new(ApiErrorCode.CertificatePinFailure, message, detail);

    public static ApiError BadRequest(string message, string? detail = null) =>
        new(ApiErrorCode.BadRequest, message, detail);

    public static ApiError NotFound(string message, string? detail = null) =>
        new(ApiErrorCode.NotFound, message, detail);

    public static ApiError Server(string message, string? detail = null) =>
        new(ApiErrorCode.ServerError, message, detail);

    public static ApiError Timeout(string message, string? detail = null) =>
        new(ApiErrorCode.Timeout, message, detail);

    public static ApiError Cancelled(string message, string? detail = null) =>
        new(ApiErrorCode.Cancelled, message, detail);

    public static ApiError Internal(string message, string? detail = null) =>
        new(ApiErrorCode.Internal, message, detail);
}

public enum ApiErrorCode
{
    Network,
    Timeout,
    Unauthorized,
    CertificatePinFailure,
    BadRequest,
    NotFound,
    ServerError,
    Cancelled,
    Internal,
}

File: native-apps/desktop/macos/Domain/Primitives/ApiError.swift
import Foundation

public enum ApiErrorCode: String, Sendable {
    case network
    case timeout
    case unauthorized
    case certificatePinFailure = "certificate_pin_failure"
    case badRequest = "bad_request"
    case notFound = "not_found"
    case serverError = "server_error"
    case cancelled
    case `internal` = "internal"
}

public struct ApiError: Error, Sendable, Equatable {
    public let code: ApiErrorCode
    public let message: String
    public let detail: String?

    public init(code: ApiErrorCode, message: String, detail: String? = nil) {
        self.code = code
        self.message = message
        self.detail = detail
    }

    public static func network(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .network, message: message, detail: detail)
    }
    public static func timeout(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .timeout, message: message, detail: detail)
    }
    public static func unauthorized(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .unauthorized, message: message, detail: detail)
    }
    public static func certificatePinFailure(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .certificatePinFailure, message: message, detail: detail)
    }
    public static func badRequest(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .badRequest, message: message, detail: detail)
    }
    public static func notFound(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .notFound, message: message, detail: detail)
    }
    public static func server(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .serverError, message: message, detail: detail)
    }
    public static func cancelled(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .cancelled, message: message, detail: detail)
    }
    public static func `internal`(_ message: String, detail: String? = nil) -> ApiError {
        .init(code: .internal, message: message, detail: detail)
    }
}

extension ApiError: LocalizedError {
    public var errorDescription: String? { message }
}
